Automattic/wp-calypso
# Inline Help Search Accessibility

### Overview

The search input will operate independently of the results. There won't be any arrow key interactions on the results while focused on the search input. 

### Markup General Structure

The previous markup was one list with `<li><h2>Section Title</h2></li>` with a combo-box-like interaction using arrow keys to navigate results. This PR changes it to a list of links with titles outside of the lists, using `TAB` to navigate the search results.

```html
<div aria-live="polite">
    
    
    <div aria-label="Search Results">
        <h3 id="section-1">Support articles</h3>
        <ul aria-labelleddby="section-1">
            <li><a href="#">Item 1</a></li>
            <li><a href="#">Item 1</a></li>
        </ul>
        <h3 id="section-2">Show me where to</h3>
        <ul role="grid" aria-labelledby="section-2">
            <li><a href="#">Item 3</a></li>
            <li><a href="#">Item 4</a></li>
        </ul>
    </div>
</div>
```

### Keyboard interactions

Everything gets a `tab`! This does add more tabstops to the page, but overall it is a more consistent and, likely, usable and standardized experience. Using `grid` roles to implement arrowed results [would likely be a confusing experience](https://adrianroselli.com/2020/07/aria-grid-as-an-anti-pattern.html) and, in the end, only benefit advanced sighted keyboard users.

### Screen Reader Announcements
I would have liked to have used `speak()` to announce state changes on the search, such as:
  - Loading/Searching 
  - New results loaded
  - Errors

There isn't an a11y `speak()` utility in Calypso and I was hesitant to add one. I opted to wrap anything loaded from the search (error message, placeholder search bars, results) in an `aria-live` region. It works fairly well, but sometimes the results are not announced. This may be due to my inexperience with a screen reader though, as I believe the implementation is correct.



#### Testing instructions

Visually, nothing should have changed. Manually test searches (including searches that would not return results, like `asdfadfasdfa`) for:
- `/home` search section
- FAB ? in the bottom right of pages

All functionality should work correctly and feel the same as on production when using a mouse.

The HTML should look similar to the HTML section in this PR, with:
- `aria-live="polite"` wrapping the results
- `<h3>Section Title</h3><ul>...</ul>` for the sections
- Placeholder lines (when searching) should be unsemantic `<div>`s instead of a `<ul>` since it wasn't a real list, but a stylized thing to show the search. 
- When searching (when placeholder search lines are showing), there should be a hidden  "Loading search results" message

If you are comfortable with VoiceOver on Safari or another screen reader, you can test the announcements by performing a search and hopefully:
- "Loading search results" is announced during a long running search
- New results are announced when a search is complete. I've found this to be inconsistent in Safari, but mostly works (it's definitely better than current state)
- Error message is announced when no results are found

Fixes https://github.com/Automattic/wp-calypso/issues/43617
